### PR TITLE
Refactor mimir.rules.kubernetes component

### PR DESCRIPTION
**Pull request by 56quarters**
#### PR Description

This change refactors the `mimir.rules.kubernetes` component to move most mutable state into a separate `eventProcessor` struct. The existing code already split most behavior between the main `rules.go` file and `events.go` file. This takes the next logical step and wraps the state from `events.go` into its own struct.

This is a prerequisite for a future PR that will add clustering support to the `mimir.rules.kubernetes` component. In the future PR a particular agent instance will be the "leader" for syncing rules from Mimir or not. When not the leader, the `eventProcesor` will be a no-op.

#### Which issue(s) this PR fixes

N/A

#### Notes to the Reviewer

As discussed offline, this change and the future PR to add clustering support is a stop-gap until there's some sort of universal clustered mode support for all components. 

This has been tested with existing unit tests and locally by running Agents and Mimir in a k3s cluster via the Helm chart.

#### PR Checklist

- [ ] CHANGELOG.md updated
- [ ] Documentation added
- [X] Tests updated
- [ ] Config converters updated

### CODE REVIEW

**Comment by 56quarters** [comment]
> loki.rules.kubernetes is very similar to the mimir.rules.kubernetes component. Would it make sense to do the same changes for the loki component in this PR and keep the similarities between the two implementations? Or should it be done via a following PR?

Good point! My preference would be to:
* Make this change only to mimir.rules.kubernetes
* Make the change to mimir.rules.kubernetes to support clustering mode
* Make the same changes to loki.rules.kubernetes based on the above PRs

**Comment by wildum** [comment]
Sounds good, would you mind creating an issue to track it and link it to this PR please?

**Comment by 56quarters** [comment]
> Sounds good, would you mind creating an issue to track it and link it to this PR please?

Done in https://github.com/grafana/agent/issues/6821

File: internal/component/mimir/rules/kubernetes/events.go [hunk 1]
```diff
@@ -157,38 +201,71 @@ func convertCRDRuleGroupToRuleGroup(crd promv1.PrometheusRuleSpec) ([]rulefmt.Ru
...
 
 	// resync mimir state after applying changes
-	return c.syncMimir(ctx)
+	return e.syncMimir(ctx)
+}
+
+// getMimirState returns the cached Mimir ruler state, rule groups indexed by Mimir namespace.
+func (e *eventProcessor) getMimirState() kubernetes.RuleGroupsByNamespace {
+	e.currentStateMtx.Lock()
+	defer e.currentStateMtx.Unlock()
```
Comment (wildum) [inline internal/component/mimir/rules/kubernetes/events.go:241]: nit: could be an RLock

**Review by wildum — APPROVED** [review]
LGTM

File: internal/component/mimir/rules/kubernetes/events.go (hunk 1 above)
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/events.go]: Fixed in 39d7ee98c511772dca751a6c79594ced837a9b03

**Comment by wildum** [comment]
I moved this PR to Alloy: https://github.com/grafana/alloy/pull/158

**Pull request closed without merging**